File: lessons/dsa/fast-and-slow-pointers/examples/two_speeds_test.go
package examples

import "testing"

var readings = []int{4, 8, 15, 16, 23, 42, 9}

func rangeOf(length int) []int {
	values := make([]int, length)
	for at := range values {
		values[at] = at
	}
	return values
}

// chain builds a straight chain of length nodes, ending rather than looping.
func chain(length int) []int {
	nexts := make([]int, length)
	for at := range nexts {
		if at == length-1 {
			nexts[at] = End
		} else {
			nexts[at] = at + 1
		}
	}
	return nexts
}

// looped builds a chain whose last node points back to entrance.
func looped(length, entrance int) []int {
	nexts := chain(length)
	nexts[length-1] = entrance
	return nexts
}

func TestTheLessonReadings(t *testing.T) {
	middle, ok := MiddleByTwoSpeeds(readings)
	if !ok || middle != 3 || readings[middle] != 16 {
		t.Fatalf("expected index 3 holding 16, got %d ok=%v", middle, ok)
	}
}

func TestBothWaysOfFindingTheMiddleAgree(t *testing.T) {
	for length := 1; length < 60; length++ {
		values := rangeOf(length)
		byTwo, okTwo := MiddleByTwoSpeeds(values)
		byCount, okCount := MiddleByCounting(values)
		if byTwo != byCount || okTwo != okCount {
			t.Fatalf("length %d: two speeds %d, counting %d", length, byTwo, byCount)
		}
	}
}

// The fast position takes two steps per round and stops at the end, so the
// rounds cannot exceed half the length. Nothing walks twice.
func TestItIsOnePass(t *testing.T) {
	for length := 1; length < 60; length++ {
		if got := StepsTaken(rangeOf(length)); got > (length+1)/2 {
			t.Fatalf("length %d took %d rounds", length, got)
		}
	}
}

// A convention rather than a discovery, pinned so a caller can rely on it.
func TestAnEvenLengthReturnsTheLaterMiddle(t *testing.T) {
	if got, _ := MiddleByTwoSpeeds([]int{0, 1, 2, 3}); got != 2 {
		t.Fatalf("expected 2, got %d", got)
	}
}

func TestEmptyAndSingle(t *testing.T) {
	if _, ok := MiddleByTwoSpeeds(nil); ok {
		t.Fatal("expected no middle in an empty sequence")
	}
	if got, ok := MiddleByTwoSpeeds([]int{9}); !ok || got != 0 {
		t.Fatalf("expected index 0, got %d", got)
	}
}

func TestAStraightChainHasNoCycle(t *testing.T) {
	for length := 1; length < 30; length++ {
		if HasCycle(chain(length), 0) {
			t.Fatalf("length %d reported a cycle", length)
		}
	}
}

// The meeting point is generally not the entrance. This checks the second phase
// against chains whose entrance is known by construction.
func TestALoopingChainHasOneAndItsEntranceIsFound(t *testing.T) {
	for length := 2; length < 30; length++ {
		for entrance := 0; entrance < length-1; entrance++ {
			nexts := looped(length, entrance)
			if !HasCycle(nexts, 0) {
				t.Fatalf("length %d entering at %d: no cycle found", length, entrance)
			}
			if got, ok := CycleEntrance(nexts, 0); !ok || got != entrance {
				t.Fatalf("length %d: expected entrance %d, got %d", length, entrance, got)
			}
		}
	}
}

func TestNoEntranceWithoutACycle(t *testing.T) {
	if _, ok := CycleEntrance(chain(10), 0); ok {
		t.Fatal("expected no entrance in a straight chain")
	}
}

func TestANodePointingAtItselfIsACycle(t *testing.T) {
	if !HasCycle([]int{0}, 0) {
		t.Fatal("expected a self-reference to be a cycle")
	}
	if got, ok := CycleEntrance([]int{0}, 0); !ok || got != 0 {
		t.Fatalf("expected entrance 0, got %d", got)
	}
}

File: lessons/dsa/fast-and-slow-pointers/examples/two-speeds.test.ts
import { expect, test } from "vitest";

import {
  END,
  cycleEntrance,
  hasCycle,
  middleByCounting,
  middleByTwoSpeeds,
  stepsTaken,
} from "./two-speeds.js";

const READINGS = [4, 8, 15, 16, 23, 42, 9];

/** A straight chain of `length` nodes, ending rather than looping. */
function chain(length: number): number[] {
  return Array.from({ length }, (_, at) => (at === length - 1 ? END : at + 1));
}

/** A chain whose last node points back to `entrance`. */
function looped(length: number, entrance: number): number[] {
  return Array.from({ length }, (_, at) =>
    at === length - 1 ? entrance : at + 1,
  );
}

test("the lesson's readings", () => {
  expect({
    middle: middleByTwoSpeeds(READINGS),
    value: READINGS[3],
  }).toEqual({ middle: 3, value: 16 });
});

test("both ways of finding the middle agree", () => {
  const wrong: number[] = [];

  for (let length = 1; length < 60; length += 1) {
    const values = Array.from({ length }, (_, at) => at);
    if (middleByTwoSpeeds(values) !== middleByCounting(values)) {
      wrong.push(length);
    }
  }

  expect(wrong).toEqual([]);
});

test("it is one pass", () => {
  // The fast position takes two steps per round and stops at the end, so the
  // rounds cannot exceed half the length. Nothing walks twice.
  const over: number[] = [];

  for (let length = 1; length < 60; length += 1) {
    const values = Array.from({ length }, (_, at) => at);
    if (stepsTaken(values) > Math.floor((length + 1) / 2)) over.push(length);
  }

  expect(over).toEqual([]);
});

test("an even length returns the later middle", () => {
  // A convention rather than a discovery, pinned so a caller can rely on it.
  expect(middleByTwoSpeeds([0, 1, 2, 3])).toBe(2);
});

test("an empty sequence has no middle", () => {
  expect([middleByTwoSpeeds([]), middleByCounting([])]).toEqual([
    undefined,
    undefined,
  ]);
});

test("a single value is its own middle", () => {
  expect(middleByTwoSpeeds([9])).toBe(0);
});

test("a straight chain has no cycle", () => {
  const wrong = Array.from({ length: 29 }, (_, at) => at + 1).filter((length) =>
    hasCycle(chain(length)),
  );

  expect(wrong).toEqual([]);
});

test("a looping chain has one, and its entrance is found rather than assumed", () => {
  // The meeting point is generally not the entrance. This checks the second
  // phase against chains whose entrance is known by construction.
  const wrong: string[] = [];

  for (let length = 2; length < 30; length += 1) {
    for (let entrance = 0; entrance < length - 1; entrance += 1) {
      const nexts = looped(length, entrance);
      if (!hasCycle(nexts) || cycleEntrance(nexts) !== entrance) {
        wrong.push(`${length} entering at ${entrance}`);
      }
    }
  }

  expect(wrong).toEqual([]);
});

test("no entrance without a cycle", () => {
  expect(cycleEntrance(chain(10))).toBeUndefined();
});

test("a node pointing at itself is a cycle", () => {
  expect({ cycle: hasCycle([0]), entrance: cycleEntrance([0]) }).toEqual({
    cycle: true,
    entrance: 0,
  });
});
